src/show.jl

```diff
@@ -0,0 +1,179 @@
+"""
+const SHOW_TREE_ICONS = Ref{Bool}(true)
+
+function Base.show(io::IO, ::MIME"text/plain", obj::Union{File,Group,Dataset,Attributes,Attribute})
+    if SHOW_TREE[]
+        show_tree(io, obj)
+    else
+        show(io, obj)
+    end
+end
+
+function _tree_icon(obj)
+    if SHOW_TREE_ICONS[]
+        return obj isa Attribute ? "🏷️" :
+               obj isa Group ? "📂" :
+               obj isa Dataset ? "🔢" :
+               obj isa Datatype ? "📄" :
+               obj isa File ? "🗂️" :
+               "❓"
+    else
+        return obj isa Attribute ? "[A]" :
+               obj isa Group ? "[G]" :
+               obj isa Dataset ? "[D]" :
+               obj isa Datatype ? "[T]" :
+               obj isa File ? "[F]" :
+               "[?]"
+    end
+end
+_tree_icon(obj::Attributes) = _tree_icon(obj.parent)
+
+_tree_head(io::IO, obj) = print(io, _tree_icon(obj), " ", obj)
+_tree_head(io::IO, obj::Datatype) = print(io, _tree_icon(obj), " HDF5.Datatype: ", name(obj))
+
+function _tree_children(parent::Union{File, Group}, attributes::Bool)
+    names = keys(parent)
+    objs  = Union{Object, Attribute}[parent[n] for n in names]
+    if attributes
+        attrn = keys(attrs(parent))
+        attro = Union{Object, Attribute}[attrs(parent)[n] for n in attrn]
+        names = append!(attrn, names)
+        objs  = append!(attro, objs)
+    end
+    return (names, objs)
+end
+function _tree_children(parent::Dataset, attributes::Bool)
+    names = String[]
+    objs = Union{Object, Attribute}[parent[n] for n in names]
+    if attributes
+        attrn = keys(attrs(parent))
+        attro = Union{Object, Attribute}[attrs(parent)[n] for n in attrn]
+        names = append!(attrn, names)
+        objs  = append!(attro, objs)
+    end
+    return (names, objs)
+end
+function _tree_children(parent::Attributes, attributes::Bool)
+    names = keys(parent)
+    objs  = Union{Object,Attribute}[parent[n] for n in names]
+    return (names, objs)
+end
+function _tree_children(parent::Union{Attribute, Datatype}, attributes::Bool)
+    # TODO: add our own implementation of much of what h5lt_dtype_to_text() does?
+    return (String[], Union{Object, Attribute}[])
+end
+
+function _show_tree(io::IO, obj::Union{File,Group,Dataset,Datatype,Attributes,Attribute}, indent::String="";
+                   attributes::Bool = true)
+    isempty(indent) && _tree_head(io, obj)
+    !isvalid(obj) && return
+
+    names, children = _tree_children(obj, attributes)
+    nchildren = length(children)
+    for ii in 1:nchildren
+        name = names[ii]
+        child  = children[ii]
+
+        islast = ii == nchildren
+        icon = _tree_icon(child)
+        print(io, "\n", indent, islast ? "└─ " : "├─ ", icon, " ", name)
+
+        nextindent = indent * (islast ? "   " : "│  ")
+        _show_tree(io, child, nextindent; attributes = attributes)
+    end
+    return nothing
+end
+
+show_tree(obj; kws...) = show_tree(stdout, obj; kws...)
+function show_tree(io::IO, obj; kws...)
```

Comment:
One final comment, are we happy with the name here or should we use `tree` ? We should probably also add this to the top of the `src.jl` module name as a  `public` but module scoped function as in here https://github.com/JuliaIO/HDF5.jl/pull/724/files